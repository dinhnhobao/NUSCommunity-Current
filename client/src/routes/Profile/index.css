.profile-bar {
  background-color: #007bff;
  padding: 16px 32px;
  box-shadow: 0 0 10px 0 rgba(0, 0, 0, 0.3);
}

.profile-bar h1 {
  color: #fff;
  font-weight: bolder;
}

.profile-center {
  padding: 24px 0px;
  background-color: #f8f9fa;
  line-height: 38px;
  text-align: center;
  display: flex;
  font-weight: bolder;
  font-size: 16px;
  margin-bottom: 24px;
}

.profile-center-back {
  flex: 1;
  margin: 0 15px;
  cursor: pointer;
  border-radius: 4px;
}

.profile-center-reset {
  flex: 1;
  margin: 0 15px;
  background: #28a745;
  color: #fff;
  border-radius: 4px;
}

.profile-center-reset:hover {
  color: #fff;
}

.profile-center-del {
  flex: 1;
  margin: 0 15px;
  background: #dc3545;
  color: #fff;
  cursor: pointer;
  border-radius: 4px;
}

.profile-con {
  border: 1px solid rgba(0, 0, 0, 0.125);
  border-radius: 4px;
}

.profile-con-title {
  padding: 0 20px;
  line-height: 60px;
  font-size: 24px;
  font-weight: bold;
  border-bottom: 1px solid rgba(0, 0, 0, 0.125);
}

.profile-con-form {
  width: 100%;
  padding: 20px;
  position: relative;
}

.profile-con-form .my-editor {
  border: 1px solid rgba(0, 0, 0, 0.125);
  height: 232px;
  padding-bottom: 32px;
}

.profile-con-form .bf-content {
  height: 105px;
  padding-bottom: 0;
}

.profile-con-form-label {
  line-height: 24px;
  margin-bottom: 8px;
  font-size: 16px;
  color: #000;
}

.profile-con-form-btn {
  position: absolute;
  border-top: 1px solid #d9d9d9;
  bottom: 44px;
  width: 1056px;
  line-height: 32px;
  text-align: center;
  cursor: pointer;
  font-weight: 600;
}
