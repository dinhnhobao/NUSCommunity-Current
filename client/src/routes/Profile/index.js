import React from "react";
import { Link } from "react-router-dom";
import {
  // UserOutlined,
  ArrowLeftOutlined,
  LockOutlined,
  DeleteOutlined,
} from "@ant-design/icons";
import { Form, message, Input, Button } from "antd";
import BraftEditor from "braft-editor";
import "braft-editor/dist/index.css";
import "./index.css";
import "./index.css";

export default class Profile extends React.Component {
  formRef = React.createRef();

  constructor(props) {
    super(props)
    this.state = {
      loggedin: false,
      email:'',
      username: '',
      points: '',
      bio:'',
      message: ''
    }
    this.handleBioChange =  this.handleBioChange.bind(this)
    this.handleProfile = this.handleProfile.bind(this)
    this.deleteAccount = this.deleteAccount.bind(this)
  }

  async componentDidMount() {
    this.getProfile()
  }

  getProfile() {
    fetch('/auth/check-session', {
      credentials: 'include'
    })
    .then(res => res.json())
    .then(response => {
      if (!response.loggedin) return;
      this.setState({
        loggedin: response.loggedin,
        email: response.email,
        username: response.username,
        points: response.points,
        bio: response.bio
      })
    })
  }

  deleteAccount() {
    console.log('Account deleted successfully triggers here!')
    fetch('/auth/deleteAccount', {
      credentials: 'include'
    })
    .then(res => res.json())
    .then(response => {
      this.setState({
        message: response.message
      })
      const { history } = this.props;
      //window.location("/");
      message.success(this.state.message);
      history.push("/");
    })
  }

  handleBioChange(event) {
    this.setState({
      bio: event.target.value
    })
  }

  handleProfile(event) {
    event.preventDefault()
    this.setState({
      bio:'',
      message: null
    })
  
    const data = {
      bio: this.state.bio
    }
    //console.log(this.state.bio);
    fetch('/auth/updateProfile', {
      method: 'POST',
      body: JSON.stringify(data),
      headers:{
        'Content-Type': 'application/json'
      }
    })
    .then(async res => {
      if (res.status === 200) {
        this.getProfile()
        this.setState({
          message: 'Profile have been saved!'
        })
        message.success(this.state.message);
      } else {
        this.setState({
          message: 'Failed to save profile. Please try again!'
        })
        message.error(this.state.message)
      }
    })
  }

  // state = {
  //   info: {
  //     name: "abc",
  //     email: "e12398132@u.nus.edu",
  //     bio: "",
  //   },
  // };

  onFinish = (values) => {
    // console.log(values);
  };

  render() {
    const { history } = this.props;
    //const { info } = this.state;
    return (
      <div className="profile">
        <div className="profile-center">
          <div
            onClick={() => {
              history.goBack();
            }}
            className="profile-center-back"
          >
            <ArrowLeftOutlined className="mr-8" />
            Back
          </div>
          <Link className="profile-center-reset" to="/reset">
            <LockOutlined className="mr-8" />
            Change Password
          </Link>
          <div className="profile-center-del" onClick={this.deleteAccount}>
            <DeleteOutlined className="mr-8" />
            Delete Account
          </div>
        </div>
        <div className="profile-con">
          <div className="profile-con-title">Edit Profile</div>
          <Form
            name="normal_add"
            ref={this.formRef}
            onFinish={this.onFinish}
            className="profile-con-form"
            initialValues={{
              bio: BraftEditor.createEditorState(this.state.bio),
            }}
          >
            <div className="profile-con-form-label">Name</div>
            <Form.Item>{this.state.username}</Form.Item>
            <div className="profile-con-form-label">Email</div>
            <Form.Item>{this.state.email}</Form.Item>
            <div className="profile-con-form-label">Contribution Points</div>
            <Form.Item>{this.state.points}</Form.Item>
            <div className="profile-con-form-label">Bio</div>
            <Form.Item>
              <Input type="text" placeholder="Bio" name="bio" id="bio" value={this.state.bio} onChange={this.handleBioChange}></Input>
            </Form.Item>
            {/* <Form.Item
              name="bio"
              validateTrigger="onBlur"
              rules={[
                {
                  validator(rule, value) {
                    if (value.isEmpty()) {
                      return Promise.reject("please enter your summary");
                    }
                    return Promise.resolve();
                  },
                },
              ]}
            >
              <BraftEditor language="en" className="my-editor" />
            </Form.Item> */}
            {/* <div
              onClick={() => {
                this.formRef.current.submit();
              }}
              className="profile-con-form-btn"
            >
              Save
            </div> */}
            <Button
              type="submit"
              onClick={this.handleProfile}
              size="large"
            >
              Save
            </Button>
          </Form>
        </div>
      </div>
    );
  }
}
