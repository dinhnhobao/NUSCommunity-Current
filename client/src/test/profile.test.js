import React from 'react';
import {shallow} from 'enzyme';
import Profile from '../routes/Profile/index';

const setUp = (props={}) => {
    const component = shallow(<Profile {...props} />);
    return component;
}

const findByTestAttr = (component, attr) => {
    const wrapper = component.find(`[data-test='${attr}']`);
    return wrapper;
}

describe('Profile Component', () => {
    let component;
    beforeEach(() => {
        component = setUp();
    })

    it('Should render correctly', () => {
        const wrapper =  findByTestAttr(component, 'profileComponent');
        expect(wrapper).toMatchSnapshot();
    });

    it('Should render without errors', () => {
        const wrapper =  findByTestAttr(component, 'profileComponent');
        expect(wrapper.length).toEqual(1);       
    });

    describe('Form', () => {
        it('Fill out bio in the form', () => {
            const wrapper = shallow(<Profile/>);
            let bioInput = wrapper.find('Input').first();
            bioInput.simulate('change', { 
                target: {value: 'This is bio section'},
            })
            bioInput = wrapper.find('Input').first();
            expect(bioInput.props().value).toEqual('This is bio section')
        })
    })

    describe('Button', () => {
        it('Should render Button with click handle', () => {
            const wrapper =  findByTestAttr(component, 'buttonComponent');
            expect(wrapper.length).toEqual(1);
        })
    })
})
