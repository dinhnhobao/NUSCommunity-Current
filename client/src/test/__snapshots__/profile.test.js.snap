// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Component Should render correctly 1`] = `ShallowWrapper {}`;
